### PR TITLE
inverse square demand following repositioning

### PR DESCRIPTION
**Pull request by haitamlaarabi**
Main changes:
+ a repositioning algorithm that breakdown the demand by hexagons (see H3TAZ)
+ uses the inverse square law to characterize demand over a distance
+ relies on a sliding window to predict future demand with predictionHorizon parameter
-----
Marginal changes:
+ addresses the issue #2515 
+ unify the computation of search distance along pooled and solo rides
+ updating naming of methods under H3TAZ and adding a new one that allows user to specify the needed resolution of the H3 index being returned.

---
This change is [Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/2529)

### CODE REVIEW

**Comment by haitamlaarabi** [comment]
Test!

**Comment by haitamlaarabi** [comment]
Test!

File: src/main/resources/beam-template.conf [hunk 1]
```diff
@@ -140,6 +141,10 @@ beam.agentsim.agents.rideHail.repositioningManager.demandFollowingRepositioningM
 beam.agentsim.agents.rideHail.repositioningManager.demandFollowingRepositioningManager.fractionOfClosestClustersToConsider = "double | 0.2"
 beam.agentsim.agents.rideHail.repositioningManager.demandFollowingRepositioningManager.horizon = "int | 1200"
 
+beam.agentsim.agents.rideHail.repositioningManager.alternativeDemandFollowingRepositioning.sensitivityOfRepositioningToDemand = "double | 0.4"
```
Comment (wrashid) [inline src/main/resources/beam-template.conf:144]: do we want to call it inverseSquareDemandFollowingRepositioning?

File: src/main/scala/beam/agentsim/agents/ridehail/repositioningmanager/InverseSquareDemandFollowingRepositioning.scala [hunk 2]
```diff
@@ -0,0 +1,200 @@
+package beam.agentsim.agents.ridehail.repositioningmanager
+
+import beam.agentsim.agents.ridehail.RideHailManager
+import beam.agentsim.agents.ridehail.RideHailVehicleManager.RideHailAgentLocation
```
Comment (wrashid) [inline src/main/scala/beam/agentsim/agents/ridehail/repositioningmanager/InverseSquareDemandFollowingRepositioning.scala:4]: should we call the file/algo: InverseSquareDistanceDemandFollowingRepositioning?

File: src/main/scala/beam/sim/config/BeamConfig.scala [hunk 3]
```diff
@@ -718,24 +718,23 @@ object BeamConfig {
...
           repositioningManager: BeamConfig.Beam.Agentsim.Agents.RideHail.RepositioningManager,
-          rideHailManager: BeamConfig.Beam.Agentsim.Agents.RideHail.RideHailManager,
           surgePricing: BeamConfig.Beam.Agentsim.Agents.RideHail.SurgePricing
         )
 
         object RideHail {
           case class AllocationManager(
             alonsoMora: BeamConfig.Beam.Agentsim.Agents.RideHail.AllocationManager.AlonsoMora,
+            maxExcessRideTime: scala.Double,
+            maxWaitingTimeInSec: scala.Int,
```
Comment (wrashid) [inline src/main/scala/beam/sim/config/BeamConfig.scala:728]: why no change to beam-template.conf?

File: test/input/beamville/beam.conf [hunk 4]
```diff
@@ -146,7 +146,7 @@ beam.agentsim.agents.rideHail.surgePricing.priceAdjustmentStrategy="KEEP_PRICE_L
 # allocationManager(DEFAULT_MANAGER | EV_MANAGER | POOLING_ALONSO_MORA)
 beam.agentsim.agents.rideHail.allocationManager.name="POOLING_ALONSO_MORA"
 beam.agentsim.agents.rideHail.allocationManager.requestBufferTimeoutInSeconds = 200
-# repositioningManager can be DEFAULT_REPOSITIONING_MANAGER | DEMAND_FOLLOWING_REPOSITIONING_MANAGER | REPOSITIONING_LOW_WAITING_TIMES
+# repositioningManager can be DEFAULT_REPOSITIONING_MANAGER | DEMAND_FOLLOWING_REPOSITIONING_MANAGER | REPOSITIONING_LOW_WAITING_TIMES | INVERSE_SQUARE_DEMAND_FOLLOWING_REPOSITIONING
```
Comment (wrashid) [inline test/input/beamville/beam.conf:149]: INVERSE_SQUARE_DISTANCE_DEMAND_FOLLOWING_REPOSITIONING

File: src/main/scala/beam/sim/config/BeamConfig.scala (hunk 3 above)
Comment (haitamlaarabi) [reply src/main/scala/beam/sim/config/BeamConfig.scala:728]: it's there

File: src/main/scala/beam/agentsim/agents/ridehail/repositioningmanager/InverseSquareDemandFollowingRepositioning.scala (hunk 2 above)
Comment (haitamlaarabi) [reply src/main/scala/beam/agentsim/agents/ridehail/repositioningmanager/InverseSquareDemandFollowingRepositioning.scala:4]: ok

**Comment by haitamlaarabi** [comment]
Test!

**Comment by haitamlaarabi** [comment]
I have renamed the three matching algorithms so that the naming reflects their functions. 
```
AlonsoMoraMatchingWithMIPAssignment
AlonsoMoraMatchingWithAsyncGreedyAssignment
AsyncGreedyVehicleCentricMatching
```

**Comment by haitamlaarabi** [comment]
Test!

**Comment by haitamlaarabi** [comment]
Test!

**Pull request merged**